File: bindings/java/src/main/java/com/dyber/quac100/Kem.java
/*
 * Copyright © 2025 Dyber, Inc. All Rights Reserved.
 * QUAC 100 Post-Quantum Cryptographic Accelerator - Java SDK
 */
package com.dyber.quac100;

/**
 * Key Encapsulation Mechanism (KEM) operations.
 * 
 * <p>
 * Provides ML-KEM (Kyber) post-quantum key exchange functionality.
 * ML-KEM is NIST's standardized lattice-based key encapsulation mechanism.
 * </p>
 * 
 * <h2>Example Usage:</h2>
 * 
 * <pre>{@code
 * Device device = lib.openFirstDevice();
 * Kem kem = device.kem();
 * 
 * // Get algorithm parameters
 * KemParams params = kem.getParams(KemAlgorithm.ML_KEM_768);
 * System.out.println("Public key size: " + params.getPublicKeySize());
 * 
 * // Alice generates a key pair
 * KeyPair aliceKeys = kem.generateKeyPair(KemAlgorithm.ML_KEM_768);
 * 
 * // Alice sends her public key to Bob...
 * 
 * // Bob encapsulates a shared secret
 * EncapsulationResult bobResult = kem.encapsulate(
 *         KemAlgorithm.ML_KEM_768, aliceKeys.getPublicKey());
 * 
 * // Bob sends ciphertext to Alice...
 * 
 * // Alice decapsulates to get the same shared secret
 * byte[] aliceSecret = kem.decapsulate(
 *         KemAlgorithm.ML_KEM_768, aliceKeys.getSecretKey(), bobResult.getCiphertext());
 * 
 * // Both parties now have the same 32-byte shared secret
 * assert Arrays.equals(bobResult.getSharedSecret(), aliceSecret);
 * }</pre>
 * 
 * @author Dyber, Inc.
 * @version 1.0.0
 */
public class Kem {

    private final Device device;

    // Native methods
    private static native KemParams nativeGetParams(long handle, int algorithm);

    private static native KeyPair nativeGenerateKeyPair(long handle, int algorithm);

    private static native EncapsulationResult nativeEncapsulate(long handle, int algorithm, byte[] publicKey);

    private static native byte[] nativeDecapsulate(long handle, int algorithm, byte[] secretKey, byte[] ciphertext);

    /**
     * Package-private constructor - instances are created by Device.
     */
    Kem(Device device) {
        this.device = device;
    }

    /**
     * Gets algorithm parameters for the specified KEM algorithm.
     * 
     * @param algorithm the KEM algorithm
     * @return KemParams containing key sizes, ciphertext size, etc.
     * @throws QuacException if operation fails
     */
    public KemParams getParams(KemAlgorithm algorithm) throws QuacException {
        KemParams params = nativeGetParams(device.getHandle(), algorithm.getValue());
        if (params == null) {
            throw new CryptoException(ErrorCode.INVALID_ALGORITHM,
                    "Failed to get parameters for " + algorithm);
        }
        return params;
    }

    /**
     * Generates a new key pair for the specified algorithm.
     * 
     * @param algorithm the KEM algorithm
     * @return KeyPair containing public and secret keys
     * @throws QuacException if key generation fails
     */
    public KeyPair generateKeyPair(KemAlgorithm algorithm) throws QuacException {
        KeyPair kp = nativeGenerateKeyPair(device.getHandle(), algorithm.getValue());
        if (kp == null) {
            throw new CryptoException(ErrorCode.CRYPTO_ERROR,
                    "Failed to generate key pair for " + algorithm);
        }
        return kp;
    }

    /**
     * Generates a new ML-KEM-512 key pair.
     * 
     * @return KeyPair containing public and secret keys
     * @throws QuacException if key generation fails
     */
    public KeyPair generateKeyPair512() throws QuacException {
        return generateKeyPair(KemAlgorithm.ML_KEM_512);
    }

    /**
     * Generates a new ML-KEM-768 key pair.
     * 
     * @return KeyPair containing public and secret keys
     * @throws QuacException if key generation fails
     */
    public KeyPair generateKeyPair768() throws QuacException {
        return generateKeyPair(KemAlgorithm.ML_KEM_768);
    }

    /**
     * Generates a new ML-KEM-1024 key pair.
     * 
     * @return KeyPair containing public and secret keys
     * @throws QuacException if key generation fails
     */
    public KeyPair generateKeyPair1024() throws QuacException {
        return generateKeyPair(KemAlgorithm.ML_KEM_1024);
    }

    /**
     * Encapsulates a shared secret using the recipient's public key.
     * 
     * @param algorithm the KEM algorithm
     * @param publicKey recipient's public key
     * @return EncapsulationResult containing ciphertext and shared secret
     * @throws QuacException if encapsulation fails
     */
    public EncapsulationResult encapsulate(KemAlgorithm algorithm, byte[] publicKey)
            throws QuacException {
        if (publicKey == null || publicKey.length == 0) {
            throw new IllegalArgumentException("Public key cannot be null or empty");
        }

        EncapsulationResult result = nativeEncapsulate(
                device.getHandle(), algorithm.getValue(), publicKey);
        if (result == null) {
            throw new CryptoException(ErrorCode.CRYPTO_ERROR,
                    "Encapsulation failed for " + algorithm);
        }
        return result;
    }

    /**
     * Encapsulates using ML-KEM-512.
     * 
     * @param publicKey recipient's public key
     * @return EncapsulationResult containing ciphertext and shared secret
     * @throws QuacException if encapsulation fails
     */
    public EncapsulationResult encapsulate512(byte[] publicKey) throws QuacException {
        return encapsulate(KemAlgorithm.ML_KEM_512, publicKey);
    }

    /**
     * Encapsulates using ML-KEM-768.
     * 
     * @param publicKey recipient's public key
     * @return EncapsulationResult containing ciphertext and shared secret
     * @throws QuacException if encapsulation fails
     */
    public EncapsulationResult encapsulate768(byte[] publicKey) throws QuacException {
        return encapsulate(KemAlgorithm.ML_KEM_768, publicKey);
    }

    /**
     * Encapsulates using ML-KEM-1024.
     * 
     * @param publicKey recipient's public key
     * @return EncapsulationResult containing ciphertext and shared secret
     * @throws QuacException if encapsulation fails
     */
    public EncapsulationResult encapsulate1024(byte[] publicKey) throws QuacException {
        return encapsulate(KemAlgorithm.ML_KEM_1024, publicKey);
    }

    /**
     * Decapsulates to recover the shared secret.
     * 
     * @param algorithm  the KEM algorithm
     * @param secretKey  recipient's secret key
     * @param ciphertext the encapsulated ciphertext
     * @return shared secret bytes
     * @throws QuacException if decapsulation fails
     */
    public byte[] decapsulate(KemAlgorithm algorithm, byte[] secretKey, byte[] ciphertext)
            throws QuacException {
        if (secretKey == null || secretKey.length == 0) {
            throw new IllegalArgumentException("Secret key cannot be null or empty");
        }
        if (ciphertext == null || ciphertext.length == 0) {
            throw new IllegalArgumentException("Ciphertext cannot be null or empty");
        }

        byte[] secret = nativeDecapsulate(
                device.getHandle(), algorithm.getValue(), secretKey, ciphertext);
        if (secret == null) {
            throw new CryptoException(ErrorCode.CRYPTO_ERROR,
                    "Decapsulation failed for " + algorithm);
        }
        return secret;
    }

    /**
     * Decapsulates using ML-KEM-512.
     * 
     * @param secretKey  recipient's secret key
     * @param ciphertext the encapsulated ciphertext
     * @return shared secret bytes
     * @throws QuacException if decapsulation fails
     */
    public byte[] decapsulate512(byte[] secretKey, byte[] ciphertext) throws QuacException {
        return decapsulate(KemAlgorithm.ML_KEM_512, secretKey, ciphertext);
    }

    /**
     * Decapsulates using ML-KEM-768.
     * 
     * @param secretKey  recipient's secret key
     * @param ciphertext the encapsulated ciphertext
     * @return shared secret bytes
     * @throws QuacException if decapsulation fails
     */
    public byte[] decapsulate768(byte[] secretKey, byte[] ciphertext) throws QuacException {
        return decapsulate(KemAlgorithm.ML_KEM_768, secretKey, ciphertext);
    }

    /**
     * Decapsulates using ML-KEM-1024.
     * 
     * @param secretKey  recipient's secret key
     * @param ciphertext the encapsulated ciphertext
     * @return shared secret bytes
     * @throws QuacException if decapsulation fails
     */
    public byte[] decapsulate1024(byte[] secretKey, byte[] ciphertext) throws QuacException {
        return decapsulate(KemAlgorithm.ML_KEM_1024, secretKey, ciphertext);
    }
}

File: bindings/nodejs/src/kem.ts
/**
 * Key Encapsulation Mechanism operations
 * @module kem
 */

import { getLib, createSizeT, readSizeT } from './ffi';
import { checkError } from './errors';
import { KemAlgorithm, KemKeyPair, EncapsulationResult, KEM_SIZES } from './types';
import type { Device } from './device';

/**
 * KEM operations interface
 *
 * Provides ML-KEM (Kyber) key encapsulation operations.
 *
 * @example
 * ```typescript
 * const kem = device.kem;
 *
 * // Generate key pair
 * const keypair = kem.generateKeypair(KemAlgorithm.MlKem768);
 *
 * // Encapsulate (sender)
 * const { ciphertext, sharedSecret } = kem.encapsulate(keypair.publicKey, KemAlgorithm.MlKem768);
 *
 * // Decapsulate (receiver)
 * const decapSecret = kem.decapsulate(keypair.secretKey, ciphertext, KemAlgorithm.MlKem768);
 *
 * // sharedSecret and decapSecret are equal
 * ```
 */
export class Kem {
    private readonly device: Device;

    constructor(device: Device) {
        this.device = device;
    }

    /**
     * Generate a KEM key pair
     * @param algorithm - KEM algorithm to use
     * @returns Key pair containing public and secret keys
     */
    generateKeypair(algorithm: KemAlgorithm): KemKeyPair {
        const lib = getLib();
        const handle = this.device.getHandle();
        const sizes = KEM_SIZES[algorithm];

        const publicKey = Buffer.alloc(sizes.publicKey);
        const secretKey = Buffer.alloc(sizes.secretKey);
        const publicKeyLen = createSizeT(sizes.publicKey);
        const secretKeyLen = createSizeT(sizes.secretKey);

        const result = lib.quac_kem_keygen(
            handle,
            algorithm,
            publicKey,
            publicKeyLen,
            secretKey,
            secretKeyLen
        );
        checkError(result);

        return {
            publicKey: publicKey.subarray(0, readSizeT(publicKeyLen)),
            secretKey: secretKey.subarray(0, readSizeT(secretKeyLen)),
        };
    }

    /**
     * Generate ML-KEM-512 key pair
     */
    generateKeypair512(): KemKeyPair {
        return this.generateKeypair(KemAlgorithm.MlKem512);
    }

    /**
     * Generate ML-KEM-768 key pair
     */
    generateKeypair768(): KemKeyPair {
        return this.generateKeypair(KemAlgorithm.MlKem768);
    }

    /**
     * Generate ML-KEM-1024 key pair
     */
    generateKeypair1024(): KemKeyPair {
        return this.generateKeypair(KemAlgorithm.MlKem1024);
    }

    /**
     * Encapsulate a shared secret using a public key
     * @param publicKey - Recipient's public key
     * @param algorithm - KEM algorithm to use
     * @returns Ciphertext and shared secret
     */
    encapsulate(publicKey: Buffer, algorithm: KemAlgorithm): EncapsulationResult {
        const lib = getLib();
        const handle = this.device.getHandle();
        const sizes = KEM_SIZES[algorithm];

        const ciphertext = Buffer.alloc(sizes.ciphertext);
        const sharedSecret = Buffer.alloc(sizes.sharedSecret);
        const ciphertextLen = createSizeT(sizes.ciphertext);
        const sharedSecretLen = createSizeT(sizes.sharedSecret);

        const result = lib.quac_kem_encapsulate(
            handle,
            algorithm,
            publicKey,
            publicKey.length,
            ciphertext,
            ciphertextLen,
            sharedSecret,
            sharedSecretLen
        );
        checkError(result);

        return {
            ciphertext: ciphertext.subarray(0, readSizeT(ciphertextLen)),
            sharedSecret: sharedSecret.subarray(0, readSizeT(sharedSecretLen)),
        };
    }

    /**
     * Encapsulate using ML-KEM-512
     */
    encapsulate512(publicKey: Buffer): EncapsulationResult {
        return this.encapsulate(publicKey, KemAlgorithm.MlKem512);
    }

    /**
     * Encapsulate using ML-KEM-768
     */
    encapsulate768(publicKey: Buffer): EncapsulationResult {
        return this.encapsulate(publicKey, KemAlgorithm.MlKem768);
    }

    /**
     * Encapsulate using ML-KEM-1024
     */
    encapsulate1024(publicKey: Buffer): EncapsulationResult {
        return this.encapsulate(publicKey, KemAlgorithm.MlKem1024);
    }

    /**
     * Decapsulate a ciphertext using a secret key
     * @param secretKey - Recipient's secret key
     * @param ciphertext - Ciphertext from encapsulation
     * @param algorithm - KEM algorithm to use
     * @returns Shared secret
     */
    decapsulate(secretKey: Buffer, ciphertext: Buffer, algorithm: KemAlgorithm): Buffer {
        const lib = getLib();
        const handle = this.device.getHandle();
        const sizes = KEM_SIZES[algorithm];

        const sharedSecret = Buffer.alloc(sizes.sharedSecret);
        const sharedSecretLen = createSizeT(sizes.sharedSecret);

        const result = lib.quac_kem_decapsulate(
            handle,
            algorithm,
            secretKey,
            secretKey.length,
            ciphertext,
            ciphertext.length,
            sharedSecret,
            sharedSecretLen
        );
        checkError(result);

        return sharedSecret.subarray(0, readSizeT(sharedSecretLen));
    }

    /**
     * Decapsulate using ML-KEM-512
     */
    decapsulate512(secretKey: Buffer, ciphertext: Buffer): Buffer {
        return this.decapsulate(secretKey, ciphertext, KemAlgorithm.MlKem512);
    }

    /**
     * Decapsulate using ML-KEM-768
     */
    decapsulate768(secretKey: Buffer, ciphertext: Buffer): Buffer {
        return this.decapsulate(secretKey, ciphertext, KemAlgorithm.MlKem768);
    }

    /**
     * Decapsulate using ML-KEM-1024
     */
    decapsulate1024(secretKey: Buffer, ciphertext: Buffer): Buffer {
        return this.decapsulate(secretKey, ciphertext, KemAlgorithm.MlKem1024);
    }

    /**
     * Get the public key size for an algorithm
     */
    getPublicKeySize(algorithm: KemAlgorithm): number {
        return KEM_SIZES[algorithm].publicKey;
    }

    /**
     * Get the secret key size for an algorithm
     */
    getSecretKeySize(algorithm: KemAlgorithm): number {
        return KEM_SIZES[algorithm].secretKey;
    }

    /**
     * Get the ciphertext size for an algorithm
     */
    getCiphertextSize(algorithm: KemAlgorithm): number {
        return KEM_SIZES[algorithm].ciphertext;
    }

    /**
     * Get the shared secret size (always 32 bytes)
     */
    getSharedSecretSize(): number {
        return 32;
    }
}